\documentclass[letterpaper]{book}
\usepackage[times,inconsolata,hyper]{Rd}
\usepackage{makeidx}
\usepackage[utf8]{inputenc} % @SET ENCODING@
% \usepackage{graphicx} % @USE GRAPHICX@
\makeindex{}
\begin{document}
\chapter*{}
\begin{center}
{\textbf{\huge Package `FIREVAT'}}
\par\bigskip{\large \today}
\end{center}
\begin{description}
\raggedright{}
\inputencoding{utf8}
\item[Type]\AsIs{Package}
\item[Title]\AsIs{FIREVAT, FInding REliable Variants without ArTifacts}
\item[Description]\AsIs{FIREVAT is a variant filtering tool for cancer sequencing data,
which uses mutational signatures to identify sequencing artifacts and
low-quality variants.}
\item[Version]\AsIs{0.5.5}
\item[Authors]\AsIs{Andy Jinseok Lee, Hyunbin Kim}
\item[Maintainer]\AsIs{Andy Jinseok Lee }\email{jinseok.lee@ncc.re.kr}\AsIs{, Hyunbin Kim }\email{khb7840@ncc.re.kr}\AsIs{}
\item[Depends]\AsIs{R (>= 3.5.0)}
\item[Imports]\AsIs{data.table,
rngtools,
doRNG,
stringi,
bedr,
GA,
jsonlite,
yaml,
lsa,
ggpubr,
caTools,
ggrepel,
gridExtra,
ggplot2,
rmarkdown,
gtable,
dplyr,
extrafont,
IRanges,
BSgenome.Hsapiens.UCSC.hg19,
BSgenome.Hsapiens.UCSC.hg38,
MutationalPatterns,
deconstructSigs}
\item[biocViews]\AsIs{CancerGenomics,
VariantRefinement,
VariantFiltering,
MutationalSignatures,
SomaticMutation,}
\item[URL]\AsIs{}\url{https://github.com/cgab-ncc/FIREVAT}\AsIs{}
\item[Encoding]\AsIs{UTF-8}
\item[LazyData]\AsIs{true}
\item[RoxygenNote]\AsIs{6.1.1}
\item[Suggests]\AsIs{knitr}
\item[VignetteBuilder]\AsIs{knitr}
\item[License]\AsIs{MIT + file LICENSE}
\end{description}
\Rdcontents{\R{} topics documented:}
\inputencoding{utf8}
\HeaderA{AnnotateVCFObj}{AnnotateVCFObj}{AnnotateVCFObj}
%
\begin{Description}\relax
Annotates a vcf.obj using df.variants.of.interest (from \code{\LinkA{PrepareAnnotationDB}{PrepareAnnotationDB}})
\end{Description}
%
\begin{Usage}
\begin{verbatim}
AnnotateVCFObj(vcf.obj, df.annotation.db, columns.to.include,
  include.all.columns = FALSE)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{vcf.obj}] \code{\LinkA{ReadVCF}{ReadVCF}}

\item[\code{df.annotation.db}] A data.frame from \code{\LinkA{PrepareAnnotationDB}{PrepareAnnotationDB}}.
This data.frame must have the columns 'CHROM', 'POS', 'REF', 'ALT'

\item[\code{columns.to.include}] A character vector of columns to include.
Note that existing columns in vcf.obj will not be affected.

\item[\code{include.all.columns}] A boolean value. If TRUE, then annotates vcf.obj with
all columns present in df.variants.of.interest. If FALSE, columns.to.include must be
supplied.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
An annotated vcf.obj
\end{Value}
\inputencoding{utf8}
\HeaderA{CheckIfVariantRefinementIsNecessary}{CheckIfVariantRefinementIsNecessary}{CheckIfVariantRefinementIsNecessary}
%
\begin{Description}\relax
Checks if variant refinement is necessary by identifying mutational signatures
related to sequencing artifact in the vcf.obj (set of original unrefined point mutations).
\end{Description}
%
\begin{Usage}
\begin{verbatim}
CheckIfVariantRefinementIsNecessary(vcf.obj, bsg, df.mut.pat.ref.sigs,
  target.mut.sigs, sequencing.artifact.mut.sigs,
  init.artifact.stop = 0.05, verbose = TRUE)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{vcf.obj}] A list from ReadVCF

\item[\code{bsg}] BSgenome.Hsapiens.UCSC object

\item[\code{df.mut.pat.ref.sigs}] A data.frame from MutPatParseRefMutSigs

\item[\code{target.mut.sigs}] A character vector of target mutational signatures from reference mutational signatures.

\item[\code{sequencing.artifact.mut.sigs}] A character vector of sequencing artifact mutational signatures from reference mutational signatures.

\item[\code{init.artifact.stop}] Numeric value less than 1. If the sum of sequencing artifact weights in vcf.obj is less than or equal to this value then
this function returns judgment = FALSE, otherwise returns judgment = TRUE.

\item[\code{verbose}] If TRUE, provides process detail. Default value is TRUE.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A list with the following elements
\begin{itemize}

\item{} judgmentA boolean value
\item{} seq.art.sigs.weights.sumA numeric value. Sum of sequencing artifact weights.

\end{itemize}

\end{Value}
\inputencoding{utf8}
\HeaderA{Chromosome.Names}{Constant}{Chromosome.Names}
\keyword{datasets}{Chromosome.Names}
%
\begin{Description}\relax
Chromosome names for FIREVAT.
Chromosome names should be given in the format of "chr" + chromosome number.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
Chromosome.Names
\end{verbatim}
\end{Usage}
%
\begin{Format}
An object of class \code{character} of length 25.
\end{Format}
\inputencoding{utf8}
\HeaderA{ComputeZScore}{ComputeZScore}{ComputeZScore}
%
\begin{Description}\relax
Returns a z-score of x given a distribution of values
\end{Description}
%
\begin{Usage}
\begin{verbatim}
ComputeZScore(values, x)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{values}] a numeric vector

\item[\code{x}] a numeric value
\end{ldescription}
\end{Arguments}
%
\begin{Value}
a numeric value corresponding to the z-score of x
\end{Value}
\inputencoding{utf8}
\HeaderA{ComputeZScoreEquiValue}{ComputeZScoreEquiValue}{ComputeZScoreEquiValue}
%
\begin{Description}\relax
Returns a numeric value that is equivalent to the specified z.score
in the distribution of 'values'
\end{Description}
%
\begin{Usage}
\begin{verbatim}
ComputeZScoreEquiValue(z.score, values)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{z.score}] numeric value

\item[\code{values}] numeric vector
\end{ldescription}
\end{Arguments}
%
\begin{Value}
a numeric value corresponding to the specified z.score in the 'values' distribution
\end{Value}
\inputencoding{utf8}
\HeaderA{DecimalCeiling}{DecimalCeiling}{DecimalCeiling}
%
\begin{Description}\relax
Returns the ceiling of a decimal value
e.g. value = 0.15, decimal = 0.1 returns 0.2
\end{Description}
%
\begin{Usage}
\begin{verbatim}
DecimalCeiling(value, decimal)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{value}] numeric value (decimal)

\item[\code{decimal}] numeric value (e.g. 0.1, 0.001)
\end{ldescription}
\end{Arguments}
%
\begin{Value}
a numeric value
\end{Value}
\inputencoding{utf8}
\HeaderA{Default.Obj.Fn}{Default.Obj.Fn}{Default.Obj.Fn}
%
\begin{Description}\relax
Calculates the default objective value for FIREVAT GA optimization.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
Default.Obj.Fn(C.refined, A.refined, C.artifactual, A.artifactual)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{C.refined}] A numeric value between 0 and 1.

\item[\code{A.refined}] A numeric value between 0 and 1.

\item[\code{C.artifactual}] A numeric value between 0 and 1.

\item[\code{A.artifactual}] A numeric value between 0 and 1.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A numeric value between 0 and 1.
\end{Value}
\inputencoding{utf8}
\HeaderA{DefaultFilterToBinary}{Transform default filtering parameters to a binary vector}{DefaultFilterToBinary}
%
\begin{Description}\relax
This function transforms default filtering parameter to binary vector
which can be used as a suggested solution in GA algorithm.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
DefaultFilterToBinary(vcf.filter, params.bit.len)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{vcf.filter}] A list generated in \code{\LinkA{MakeFilter}{MakeFilter}}

\item[\code{params.bit.len}] A list with bit lengths of filtering parameters which is generated from \code{\LinkA{ParameterToBits}{ParameterToBits}}
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A binary vector
\end{Value}
\inputencoding{utf8}
\HeaderA{EnumerateTriNucCounts}{EnumerateTriNucCounts}{EnumerateTriNucCounts}
%
\begin{Description}\relax
Returns C>A, C>G, C>T, T>A, T>C, T>G counts
\end{Description}
%
\begin{Usage}
\begin{verbatim}
EnumerateTriNucCounts(spectrum)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{spectrum}] a numeric vector with 96 numeric values
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Please note that this function assumes that 'spectrum' is sorted
(i.e. 1:16  --> C>A;
17:32 --> C>G;
33:48 --> C>T;
49:64 --> T>A;
65:80 --> T>C;
81:96 --> T>G)
\end{Details}
%
\begin{Value}
a numeric vector of length 6 corresponding to the counts of each trinucleotide change (C>A, C>G, C>T, T>A, T>C, T>G)
\end{Value}
\inputencoding{utf8}
\HeaderA{Euc.Exp.Weighted.Obj.Fn}{Euc.Exp.Weighted.Obj.Fn}{Euc.Exp.Weighted.Obj.Fn}
%
\begin{Description}\relax
Calculates the Euclidean-distance of logarithmically weighted
objective value for FIREVAT GA optimization.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
Euc.Exp.Weighted.Obj.Fn(C.refined, A.refined, C.artifactual, A.artifactual)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{C.refined}] A numeric value between 0 and 1.

\item[\code{A.refined}] A numeric value between 0 and 1.

\item[\code{C.artifactual}] A numeric value between 0 and 1.

\item[\code{A.artifactual}] A numeric value between 0 and 1.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A numeric value between 0 and 1.
\end{Value}
\inputencoding{utf8}
\HeaderA{Euc.Exp.Weighted.Seq.Art.Only.Obj.Fn.1}{Euc.Exp.Weighted.Seq.Art.Only.Obj.Fn.1}{Euc.Exp.Weighted.Seq.Art.Only.Obj.Fn.1}
%
\begin{Description}\relax
Calculates the Euclidean-distance of logarithmically weighted
objective value for FIREVAT GA optimization.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
Euc.Exp.Weighted.Seq.Art.Only.Obj.Fn.1(C.refined, A.refined, C.artifactual,
  A.artifactual)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{C.refined}] A numeric value between 0 and 1.

\item[\code{A.refined}] A numeric value between 0 and 1.

\item[\code{C.artifactual}] A numeric value between 0 and 1.

\item[\code{A.artifactual}] A numeric value between 0 and 1.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A numeric value between 0 and 1.
\end{Value}
\inputencoding{utf8}
\HeaderA{Euc.Exp.Weighted.Seq.Art.Only.Obj.Fn.2}{Euc.Exp.Weighted.Seq.Art.Only.Obj.Fn.2}{Euc.Exp.Weighted.Seq.Art.Only.Obj.Fn.2}
%
\begin{Description}\relax
Calculates the Euclidean-distance of logarithmically weighted
objective value for FIREVAT GA optimization.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
Euc.Exp.Weighted.Seq.Art.Only.Obj.Fn.2(C.refined, A.refined, C.artifactual,
  A.artifactual)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{C.refined}] A numeric value between 0 and 1.

\item[\code{A.refined}] A numeric value between 0 and 1.

\item[\code{C.artifactual}] A numeric value between 0 and 1.

\item[\code{A.artifactual}] A numeric value between 0 and 1.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A numeric value between 0 and 1.
\end{Value}
\inputencoding{utf8}
\HeaderA{Euc.Obj.Fn}{Euc.Obj.Fn}{Euc.Obj.Fn}
%
\begin{Description}\relax
Calculates the Euclidean-distance based objective value for FIREVAT GA optimization.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
Euc.Obj.Fn(C.refined, A.refined, C.artifactual, A.artifactual)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{C.refined}] A numeric value between 0 and 1.

\item[\code{A.refined}] A numeric value between 0 and 1.

\item[\code{C.artifactual}] A numeric value between 0 and 1.

\item[\code{A.artifactual}] A numeric value between 0 and 1.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A numeric value between 0 and 1.
\end{Value}
\inputencoding{utf8}
\HeaderA{Exp.Weighted.A.Art.Obj.Fn}{Exp.Weighted.A.Art.Obj.Fn}{Exp.Weighted.A.Art.Obj.Fn}
%
\begin{Description}\relax
Exponentially weighted objective function
\end{Description}
%
\begin{Usage}
\begin{verbatim}
Exp.Weighted.A.Art.Obj.Fn(C.refined, A.refined, C.artifactual,
  A.artifactual)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{C.refined}] A numeric value between 0 and 1.

\item[\code{A.refined}] A numeric value between 0 and 1.

\item[\code{C.artifactual}] A numeric value between 0 and 1.

\item[\code{A.artifactual}] A numeric value between 0 and 1.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A numeric value between 0 and 1.
\end{Value}
\inputencoding{utf8}
\HeaderA{Exp.Weighted.A.Ref.Obj.Fn}{Exp.Weighted.A.Ref.Obj.Fn}{Exp.Weighted.A.Ref.Obj.Fn}
%
\begin{Description}\relax
Exponentially weighted objective function
\end{Description}
%
\begin{Usage}
\begin{verbatim}
Exp.Weighted.A.Ref.Obj.Fn(C.refined, A.refined, C.artifactual,
  A.artifactual)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{C.refined}] A numeric value between 0 and 1.

\item[\code{A.refined}] A numeric value between 0 and 1.

\item[\code{C.artifactual}] A numeric value between 0 and 1.

\item[\code{A.artifactual}] A numeric value between 0 and 1.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A numeric value between 0 and 1.
\end{Value}
\inputencoding{utf8}
\HeaderA{Exp.Weighted.Obj.Fn.1}{Exp.Weighted.Obj.Fn.1}{Exp.Weighted.Obj.Fn.1}
%
\begin{Description}\relax
Calculates the exponentially weighted objective value
for FIREVAT GA optimization.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
Exp.Weighted.Obj.Fn.1(C.refined, A.refined, C.artifactual, A.artifactual)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{C.refined}] A numeric value between 0 and 1.

\item[\code{A.refined}] A numeric value between 0 and 1.

\item[\code{C.artifactual}] A numeric value between 0 and 1.

\item[\code{A.artifactual}] A numeric value between 0 and 1.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A numeric value between 0 and 1.
\end{Value}
\inputencoding{utf8}
\HeaderA{Exp.Weighted.Obj.Fn.2}{Exp.Weighted.Obj.Fn.2}{Exp.Weighted.Obj.Fn.2}
%
\begin{Description}\relax
Calculates the exponentially weighted objective value
for FIREVAT GA optimization.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
Exp.Weighted.Obj.Fn.2(C.refined, A.refined, C.artifactual, A.artifactual)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{C.refined}] A numeric value between 0 and 1.

\item[\code{A.refined}] A numeric value between 0 and 1.

\item[\code{C.artifactual}] A numeric value between 0 and 1.

\item[\code{A.artifactual}] A numeric value between 0 and 1.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A numeric value between 0 and 1.
\end{Value}
\inputencoding{utf8}
\HeaderA{Exp.Weighted.Refined.Seq.Art.Only.Obj.Fn}{Exp.Weighted.Refined.Seq.Art.Only.Obj.Fn}{Exp.Weighted.Refined.Seq.Art.Only.Obj.Fn}
%
\begin{Description}\relax
Calculates the Euclidean-distance of logarithmically weighted
objective value for FIREVAT GA optimization.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
Exp.Weighted.Refined.Seq.Art.Only.Obj.Fn(C.refined, A.refined,
  C.artifactual, A.artifactual)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{C.refined}] A numeric value between 0 and 1.

\item[\code{A.refined}] A numeric value between 0 and 1.

\item[\code{C.artifactual}] A numeric value between 0 and 1.

\item[\code{A.artifactual}] A numeric value between 0 and 1.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A numeric value between 0 and 1.
\end{Value}
\inputencoding{utf8}
\HeaderA{FilterByStrandBiasAnalysis}{FilterByStrandBiasAnalysis}{FilterByStrandBiasAnalysis}
%
\begin{Description}\relax
Filters refined.vcf.obj by strand bias analysis and
moves these filtered variants to artifactual.vcf.obj
\end{Description}
%
\begin{Usage}
\begin{verbatim}
FilterByStrandBiasAnalysis(refined.vcf.obj, artifactual.vcf.obj,
  perform.fdr.correction, filter.by.strand.bias.analysis.cutoff)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{refined.vcf.obj}] A list of vcf data

\item[\code{artifactual.vcf.obj}] A list of vcf data

\item[\code{perform.fdr.correction}] A boolean value.

\item[\code{filter.by.strand.bias.analysis.cutoff}] A numeric value.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A list with filtering parameter values
\begin{itemize}

\item{} refined.vcf.obj updated refined.vcf.obj
\item{} artifactual.vcf.obj updated artifactual.vcf.obj

\end{itemize}

\end{Value}
\inputencoding{utf8}
\HeaderA{FilterVCF}{FilterVCF}{FilterVCF}
%
\begin{Description}\relax
Filter vcf based on the filter
Filtering parameters are saved in config.obj
Split vcf.obj into vcf.obj.filtered \& vcf.obj.artifact based on vcf.filter
\end{Description}
%
\begin{Usage}
\begin{verbatim}
FilterVCF(vcf.obj, vcf.filter, config.obj, include.array = NULL,
  force.include = FALSE, verbose = TRUE)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{vcf.obj}] A list from ReadVCF

\item[\code{vcf.filter}] A list from MakeMuTect2Filter

\item[\code{config.obj}] A list from ParseConfigFile

\item[\code{include.array}] A boolean vector

\item[\code{force.include}] A boolean value. If TRUE, then uses 'include.array'

\item[\code{verbose}] If true, provides process detail
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A list with the following elements
\begin{itemize}

\item{} 1) Mutations which passed filteringvcf.obj.filtered = vcf.obj (list with data, header, genome)
\item{} 2) Mutations which did not pass filteringvcf.obj.artifact = vcf.obj (list with data, header, genome)

\end{itemize}

\end{Value}
\inputencoding{utf8}
\HeaderA{GenerateConfigObj}{Generate config.obj by checking vcf header}{GenerateConfigObj}
%
\begin{Description}\relax
This function generate config.obj by checking vcf header.
Users should fill in the information needed in console.
In current version, only Integers \& Float values can be used in
config.obj for running FIREVAT.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
GenerateConfigObj(vcf.obj, save.config = TRUE,
  config.path = "../temp/FIREVAT_configure.json")
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{vcf.obj}] A list from \code{\LinkA{ReadVCF}{ReadVCF}}

\item[\code{save.config}] If true, save config.obj to config.path

\item[\code{config.path}] File path to write config.obj (json or yaml)
\end{ldescription}
\end{Arguments}
%
\begin{Value}
config.obj
\end{Value}
\inputencoding{utf8}
\HeaderA{GetCOSMICMutSigs}{GetCOSMICMutSigs}{GetCOSMICMutSigs}
%
\begin{Description}\relax
Returns a data.frame of the COSMIC mutational signature reference file
from the data directory
\end{Description}
%
\begin{Usage}
\begin{verbatim}
GetCOSMICMutSigs()
\end{verbatim}
\end{Usage}
%
\begin{Value}
a data.frame of the COSMIC reference mutational signatures
\end{Value}
\inputencoding{utf8}
\HeaderA{GetCOSMICMutSigsEtiologiesColors}{GetCOSMICMutSigsNames}{GetCOSMICMutSigsEtiologiesColors}
%
\begin{Description}\relax
Returns all COSMIC mutational signature etiologies and colors
\end{Description}
%
\begin{Usage}
\begin{verbatim}
GetCOSMICMutSigsEtiologiesColors()
\end{verbatim}
\end{Usage}
%
\begin{Value}
data.frame with following columns: signature, group and color.
\end{Value}
\inputencoding{utf8}
\HeaderA{GetCOSMICMutSigsNames}{GetCOSMICMutSigsNames}{GetCOSMICMutSigsNames}
%
\begin{Description}\relax
Returns all COSMIC mutational signature names
\end{Description}
%
\begin{Usage}
\begin{verbatim}
GetCOSMICMutSigsNames()
\end{verbatim}
\end{Usage}
%
\begin{Value}
a character vector
\end{Value}
\inputencoding{utf8}
\HeaderA{GetGASuggestedSolutions}{GetGASuggestedSolutions}{GetGASuggestedSolutions}
%
\begin{Description}\relax
Computes suggested solutions
\end{Description}
%
\begin{Usage}
\begin{verbatim}
GetGASuggestedSolutions(vcf.obj, bsg, config.obj, lower.upper.list,
  df.mut.pat.ref.sigs, target.mut.sigs, sequencing.artifact.mut.sigs,
  objective.fn, original.muts.seq.art.weights.sum, ga.preemptive.killing,
  verbose = TRUE)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{vcf.obj}] A list from ReadVCF

\item[\code{bsg}] BSgenome.Hsapiens.UCSC object

\item[\code{config.obj}] A list from ParseConfigFile

\item[\code{lower.upper.list}] A list from GetParameterLowerUpperVector

\item[\code{df.mut.pat.ref.sigs}] A data.frame from MutPatParseRefMutSigs

\item[\code{target.mut.sigs}] A character vector of the target mutational signatures from reference mutational signatures.

\item[\code{sequencing.artifact.mut.sigs}] A character vector of the sequencing artifact mutational signatures from reference mutational signatures.

\item[\code{objective.fn}] Objective value derivation function.

\item[\code{original.muts.seq.art.weights.sum}] A numeric value. 'seq.art.sigs.weights.sum' from CheckIfVariantRefinementIsNecessary

\item[\code{ga.preemptive.killing}] If TRUE, then preemptively kills populations that yield greater sequencing artifact weights sum
compared to the original mutatational signatures analysis

\item[\code{verbose}] If TRUE, provides process detail. Default value is TRUE.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A list with the following elements
\begin{itemize}

\item{} judgmentA boolean value
\item{} seq.art.sigs.weightsA numeric value. Sum of sequencing artifact weights.

\end{itemize}

\end{Value}
\inputencoding{utf8}
\HeaderA{GetOptimizedSignatures}{GetOptimizedSignatures}{GetOptimizedSignatures}
%
\begin{Description}\relax
This function fetches the last row from the optimization iteration log
and returns the target and artifactual mutational signatures
for the type of mutations ('refined' or 'artifactual')
\end{Description}
%
\begin{Usage}
\begin{verbatim}
GetOptimizedSignatures(data, mutations.type = "refined",
  signatures = "all")
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{data}] A list of main data from \code{\LinkA{RunFIREVAT}{RunFIREVAT}}

\item[\code{mutations.type}] A string for type of mutations ('refined' or 'artifact')

\item[\code{signatures}] A string ('all', 'target', 'artifact')
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A data.frame with the columns 'signature' and 'weight'
\end{Value}
\inputencoding{utf8}
\HeaderA{GetParameterLowerUpperVector}{GetParameterLowerUpperVector}{GetParameterLowerUpperVector}
%
\begin{Description}\relax
Return a lower/upper vector needed to conduct FIREVAT GA real-valued optimization.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
GetParameterLowerUpperVector(vcf.obj, config.obj, vcf.filter,
  multiplier = 100)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{vcf.obj}] A list from ReadVCF

\item[\code{config.obj}] A list from ParseConfigFile

\item[\code{vcf.filter}] A list from MakeMuTect2Filter

\item[\code{multiplier}] A multiplier for convert fraction to integer (default = 100)
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
vcf.obj\$data: if max(vcf.obj\$data[[param]]) < 1, then multiply multiplier to the vector
\end{Details}
%
\begin{Value}
A list with the elements
\begin{itemize}

\item{} lower.vector A numeric vector. Each element is the minimum value of each parameter
\item{} upper.vector A numeric vector. Each element is the maximum value of each parameter
\item{} vcf.obj vcf.obj with updated data

\end{itemize}

\end{Value}
\inputencoding{utf8}
\HeaderA{GetPCAWGMutSigs}{GetPCAWGMutSigs}{GetPCAWGMutSigs}
%
\begin{Description}\relax
Returns the PCAWG mutational signatures data
\end{Description}
%
\begin{Usage}
\begin{verbatim}
GetPCAWGMutSigs(sequencing.type = "wes")
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{sequencing.type}] A string value.
It can be either 'wes' for whole-exome sequencing or 'wgs' for whole-genome sequencing
\end{ldescription}
\end{Arguments}
%
\begin{Value}
a data.frame of the PCAWG mutatioanl signatures
\end{Value}
\inputencoding{utf8}
\HeaderA{GetPCAWGMutSigsEtiologiesColors}{GetPCAWGMutSigsEtiologiesColors}{GetPCAWGMutSigsEtiologiesColors}
%
\begin{Description}\relax
Returns the PCAWG mutational signatures etiologies and colors
\end{Description}
%
\begin{Usage}
\begin{verbatim}
GetPCAWGMutSigsEtiologiesColors()
\end{verbatim}
\end{Usage}
%
\begin{Value}
a data.frame with the columns 'signature', 'group', 'color'
\end{Value}
\inputencoding{utf8}
\HeaderA{GetPCAWGMutSigsNames}{GetPCAWGMutSigsNames}{GetPCAWGMutSigsNames}
%
\begin{Description}\relax
Returns the PCAWG mutational signatures names
\end{Description}
%
\begin{Usage}
\begin{verbatim}
GetPCAWGMutSigsNames()
\end{verbatim}
\end{Usage}
%
\begin{Value}
a character vector of the PCAWG mutational signatures names
\end{Value}
\inputencoding{utf8}
\HeaderA{GetPCAWGPlatinumMutSigs}{GetPCAWGPlatinumMutSigs}{GetPCAWGPlatinumMutSigs}
%
\begin{Description}\relax
Returns the PCAWG platinum mutational signatures data
\end{Description}
%
\begin{Usage}
\begin{verbatim}
GetPCAWGPlatinumMutSigs()
\end{verbatim}
\end{Usage}
%
\begin{Value}
a data.frame of the PCAWG platinum mutatioanl signatures
\end{Value}
\inputencoding{utf8}
\HeaderA{GetPCAWGPlatinumMutSigsEtiologiesColors}{GetPCAWGPlatinumMutSigsEtiologiesColors}{GetPCAWGPlatinumMutSigsEtiologiesColors}
%
\begin{Description}\relax
Returns the PCAWG platinum mutational signatures etiologies and colors
\end{Description}
%
\begin{Usage}
\begin{verbatim}
GetPCAWGPlatinumMutSigsEtiologiesColors()
\end{verbatim}
\end{Usage}
%
\begin{Value}
a data.frame with the columns 'signature', 'group', 'color'
\end{Value}
\inputencoding{utf8}
\HeaderA{GetPCAWGPlatinumMutSigsNames}{GetPCAWGPlatinumMutSigsNames}{GetPCAWGPlatinumMutSigsNames}
%
\begin{Description}\relax
Returns the PCAWG platinum mutational signatures names
\end{Description}
%
\begin{Usage}
\begin{verbatim}
GetPCAWGPlatinumMutSigsNames()
\end{verbatim}
\end{Usage}
%
\begin{Value}
a character vector of the PCAWG platinum mutational signatures names
\end{Value}
\inputencoding{utf8}
\HeaderA{GetSeedForVCF}{GetSeedForVCF}{GetSeedForVCF}
%
\begin{Description}\relax
Returns a seed integer based on VCF file size
\end{Description}
%
\begin{Usage}
\begin{verbatim}
GetSeedForVCF(vcf.file)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{vcf.file}] (full path of a .vcf file)
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
Returns the same seed integer for the same VCF file (based on file size)
\end{Details}
%
\begin{Value}
an integer value
\end{Value}
\inputencoding{utf8}
\HeaderA{InitializeVCF}{InitializeVCF}{InitializeVCF}
%
\begin{Description}\relax
Initialize VCF with FIREVAT config file
This functions selects point mutations and
appends filter values to vcf.obj\$data
\end{Description}
%
\begin{Usage}
\begin{verbatim}
InitializeVCF(vcf.obj, config.obj, verbose = TRUE)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{vcf.obj}] A list from ReadVCF

\item[\code{config.obj}] A list from ParseConfigFile

\item[\code{verbose}] If true, provides process detail
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A list with the following elements
\begin{itemize}

\item{} vcf.obj.filteredvcf.obj (high-quality vcf)
\item{} vcf.obj.artifactvcf.obj (low-quality vcf)

\end{itemize}

\end{Value}
\inputencoding{utf8}
\HeaderA{Leaky.ReLU.A.Art.Obj.Fn}{Leaky.ReLU.A.Art.Obj.Fn}{Leaky.ReLU.A.Art.Obj.Fn}
%
\begin{Description}\relax
Leaky ReLU objective function
\end{Description}
%
\begin{Usage}
\begin{verbatim}
Leaky.ReLU.A.Art.Obj.Fn(C.refined, A.refined, C.artifactual, A.artifactual)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{C.refined}] A numeric value between 0 and 1.

\item[\code{A.refined}] A numeric value between 0 and 1.

\item[\code{C.artifactual}] A numeric value between 0 and 1.

\item[\code{A.artifactual}] A numeric value between 0 and 1.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A numeric value between 0 and 1.
\end{Value}
\inputencoding{utf8}
\HeaderA{Leaky.ReLU.A.Ref.Obj.Fn}{Leaky.ReLU.A.Ref.Obj.Fn}{Leaky.ReLU.A.Ref.Obj.Fn}
%
\begin{Description}\relax
Leaky ReLU objective function
\end{Description}
%
\begin{Usage}
\begin{verbatim}
Leaky.ReLU.A.Ref.Obj.Fn(C.refined, A.refined, C.artifactual, A.artifactual)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{C.refined}] A numeric value between 0 and 1.

\item[\code{A.refined}] A numeric value between 0 and 1.

\item[\code{C.artifactual}] A numeric value between 0 and 1.

\item[\code{A.artifactual}] A numeric value between 0 and 1.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A numeric value between 0 and 1.
\end{Value}
\inputencoding{utf8}
\HeaderA{Leaky.ReLU.Obj.Fn}{Leaky.ReLU.Obj.Fn}{Leaky.ReLU.Obj.Fn}
%
\begin{Description}\relax
Lkeay ReLU objective function
\end{Description}
%
\begin{Usage}
\begin{verbatim}
Leaky.ReLU.Obj.Fn(C.refined, A.refined, C.artifactual, A.artifactual)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{C.refined}] A numeric value between 0 and 1.

\item[\code{A.refined}] A numeric value between 0 and 1.

\item[\code{C.artifactual}] A numeric value between 0 and 1.

\item[\code{A.artifactual}] A numeric value between 0 and 1.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A numeric value between 0 and 1.
\end{Value}
\inputencoding{utf8}
\HeaderA{MakeFilter}{MakeFilter}{MakeFilter}
%
\begin{Description}\relax
Creates a vcf filter from config.obj
\end{Description}
%
\begin{Usage}
\begin{verbatim}
MakeFilter(config.obj)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{config.obj}] A list from ParseConfigFile
(any filter with "use\_in\_filter" value declared as FALSE is not considered)
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A list with the filter parameters
\end{Value}
\inputencoding{utf8}
\HeaderA{MutaliskParseVCFObj}{MutaliskParseVCFObj}{MutaliskParseVCFObj}
%
\begin{Description}\relax
Parses a vcf.obj and prepares it to run Mutalisk.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
MutaliskParseVCFObj(vcf.obj)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{vcf.obj}] A list from ReadVCF
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A data.frame
\end{Value}
\inputencoding{utf8}
\HeaderA{MutPatParseRefMutSigs}{MutPatParseRefMutSigs}{MutPatParseRefMutSigs}
%
\begin{Description}\relax
Parses a df.ref.mut.sigs and prepares it to run Mutational Patterns.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
MutPatParseRefMutSigs(df.ref.mut.sigs, target.mut.sigs,
  signature.start.column.index = 4,
  mutation.type.header = "SomaticMutationType")
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{df.ref.mut.sigs}] A data.frame of reference mutational signatures

\item[\code{target.mut.sigs}] A character vector of target mutational signatures names

\item[\code{signature.start.column.index}] = An integer value (e.g. column index corresponding to 'SBS1')

\item[\code{mutation.type.header}] = A string value
(name of header corresponding to column containing 'A[C>A]A' data))
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A data.frame of the format deconstructSigs::signatures.cosmic
\end{Value}
\inputencoding{utf8}
\HeaderA{MutPatParseVCFObj}{MutPatParseVCFObj}{MutPatParseVCFObj}
%
\begin{Description}\relax
Parses a vcf.obj and prepares it to run Mutational Patterns.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
MutPatParseVCFObj(vcf.obj, bsg, sample.id = "sample")
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{vcf.obj}] A list from ReadVCF

\item[\code{bsg}] BSgenome.Hsapiens.UCSC.hg19::BSgenome.Hsapiens.UCSC.hg19 or
BSgenome.Hsapiens.UCSC.hg38::BSgenome.Hsapiens.UCSC.hg38

\item[\code{sample.id}] A string value
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A data.frame with the column sample.id and
row names corresponding to 96 substitution types
\end{Value}
\inputencoding{utf8}
\HeaderA{ParameterToBits}{ParameterToBits}{ParameterToBits}
%
\begin{Description}\relax
Calculate the number of bits needed to conduct FIREVAT GA binary optimization.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
ParameterToBits(vcf.obj, config.obj, vcf.filter, multiplier = 100)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{vcf.obj}] A list from ReadVCF

\item[\code{config.obj}] A list from ParseConfigFile

\item[\code{vcf.filter}] A list from MakeMuTect2Filter

\item[\code{multiplier}] A multiplier for convert fraction to integer (default = 100)
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
vcf.obj\$data: if max(vcf.obj\$data[[param]]) < 1, then multiply multiplier to the vector
\end{Details}
%
\begin{Value}
A list with the elements
\begin{itemize}

\item{} params.bit.lenA numeric vector. Each element is the bit length of each parameter value
\item{} vcf.objA vcf.obj (\code{\LinkA{ReadVCF}{ReadVCF}}) with updated data

\end{itemize}

\end{Value}
\inputencoding{utf8}
\HeaderA{ParseConfigFile}{ParseConfigFile}{ParseConfigFile}
%
\begin{Description}\relax
This function returns config.obj from JSON or YAML config file.
- Check if the config file is in JSON format or YAML format
- Return config.obj
\end{Description}
%
\begin{Usage}
\begin{verbatim}
ParseConfigFile(config.path, verbose = TRUE)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{config.path}] A string for config file path

\item[\code{verbose}] If true, provides process detail
\end{ldescription}
\end{Arguments}
%
\begin{Value}
config.obj: list of parameters
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
## Not run: 
ParseConfigFile("example.variant.caller.json")
ParseConfigFile("example.variant.caller.json", verbose=False)

## End(Not run)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{PCAWG.All.Sequencing.Artifact.Signatures}{Constant}{PCAWG.All.Sequencing.Artifact.Signatures}
\keyword{datasets}{PCAWG.All.Sequencing.Artifact.Signatures}
%
\begin{Description}\relax
PCAWG mutational signatures reported to be associated with sequencing artifacts
\end{Description}
%
\begin{Usage}
\begin{verbatim}
PCAWG.All.Sequencing.Artifact.Signatures
\end{verbatim}
\end{Usage}
%
\begin{Format}
An object of class \code{character} of length 18.
\end{Format}
\inputencoding{utf8}
\HeaderA{PCAWG.Known.Sequencing.Artifact.Signatures}{Constant}{PCAWG.Known.Sequencing.Artifact.Signatures}
\keyword{datasets}{PCAWG.Known.Sequencing.Artifact.Signatures}
%
\begin{Description}\relax
PCAWG mutational signatures reported to be associated with sequencing artifacts
\end{Description}
%
\begin{Usage}
\begin{verbatim}
PCAWG.Known.Sequencing.Artifact.Signatures
\end{verbatim}
\end{Usage}
%
\begin{Format}
An object of class \code{character} of length 1.
\end{Format}
\inputencoding{utf8}
\HeaderA{PCAWG.Platinum.All.Technology.Related.Artifact.Signatures}{Constant}{PCAWG.Platinum.All.Technology.Related.Artifact.Signatures}
\keyword{datasets}{PCAWG.Platinum.All.Technology.Related.Artifact.Signatures}
%
\begin{Description}\relax
PCAWG mutational signatures reported to be associated with sequencing artifacts
\end{Description}
%
\begin{Usage}
\begin{verbatim}
PCAWG.Platinum.All.Technology.Related.Artifact.Signatures
\end{verbatim}
\end{Usage}
%
\begin{Format}
An object of class \code{character} of length 9.
\end{Format}
\inputencoding{utf8}
\HeaderA{PCAWG.Possible.Sequencing.Artifact.Signatures}{Constant}{PCAWG.Possible.Sequencing.Artifact.Signatures}
\keyword{datasets}{PCAWG.Possible.Sequencing.Artifact.Signatures}
%
\begin{Description}\relax
PCAWG mutational signatures reported to be associated with sequencing artifacts
\end{Description}
%
\begin{Usage}
\begin{verbatim}
PCAWG.Possible.Sequencing.Artifact.Signatures
\end{verbatim}
\end{Usage}
%
\begin{Format}
An object of class \code{character} of length 17.
\end{Format}
\inputencoding{utf8}
\HeaderA{PCAWG.Target.Mutational.Signatures}{Constant}{PCAWG.Target.Mutational.Signatures}
\keyword{datasets}{PCAWG.Target.Mutational.Signatures}
%
\begin{Description}\relax
PCAWG target mutational signatures reported to be unrelated to sequencing artifacts
\end{Description}
%
\begin{Usage}
\begin{verbatim}
PCAWG.Target.Mutational.Signatures
\end{verbatim}
\end{Usage}
%
\begin{Format}
An object of class \code{character} of length 47.
\end{Format}
\inputencoding{utf8}
\HeaderA{PerformStrandBiasAnalysis}{PerformStrandBiasAnalysis}{PerformStrandBiasAnalysis}
%
\begin{Description}\relax
Performs strand bias analysis
\end{Description}
%
\begin{Usage}
\begin{verbatim}
PerformStrandBiasAnalysis(vcf.obj, ref.forward.strand.var,
  ref.reverse.strand.var, alt.forward.strand.var, alt.reverse.strand.var,
  perform.fdr.correction = TRUE, fdr.correction.method = "BH")
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{vcf.obj}] \code{\LinkA{ReadVCF}{ReadVCF}}

\item[\code{ref.forward.strand.var}] A string value.

\item[\code{ref.reverse.strand.var}] A string value.

\item[\code{alt.forward.strand.var}] A string value.

\item[\code{alt.reverse.strand.var}] A string value.

\item[\code{perform.fdr.correction}] If TRUE, then performs false discovery rate correction

\item[\code{fdr.correction.method}] A string value. FDR correction method (Refer to p.adjust() function)
\end{ldescription}
\end{Arguments}
%
\begin{Value}
An updated vcf.obj
\end{Value}
\inputencoding{utf8}
\HeaderA{PlotMutaliskResults}{PlotMutaliskResults}{PlotMutaliskResults}
%
\begin{Description}\relax
Plots Mutalisk results
\end{Description}
%
\begin{Usage}
\begin{verbatim}
PlotMutaliskResults(mutalisk.results, signatures,
  df.ref.sigs.groups.colors, trinuc.max.y, trinuc.min.y, mut.type.max.y,
  title, font.size.small = 8, font.size.med = 14)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{mutalisk.results}] A list obtained from \code{\LinkA{RunMutalisk}{RunMutalisk}}

\item[\code{signatures}] A character vector of mutational signatures names

\item[\code{df.ref.sigs.groups.colors}] A data.frame with signature groups and colors

\item[\code{trinuc.max.y}] A numeric value (maximum y-axis value)

\item[\code{trinuc.min.y}] A numeric value (minimum y-axis value)

\item[\code{mut.type.max.y}] A numeric value

\item[\code{title}] A string value

\item[\code{font.size.small}] A numeric value

\item[\code{font.size.med}] A numeric value
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A ggplot object
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
## Not run: 
  df.ref.mut.sigs <- GetPCAWGMutSigs()
  target.mut.sigs <- GetPCAWGMutSigsNames()
  vcf.obj <- ReadVCF(vcf.file = "../data/sample/P-233-CT.final.vcf")
  mutalisk.results <- RunMutalisk(vcf.obj = vcf.obj,
                                  df.ref.mut.sigs = df.ref.mut.sigs,
                                  target.mut.sigs = target.mut.sigs)
  p <- PlotMutaliskResults(mutalisk.results = mutalisk.results)
  print(p)

## End(Not run)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{PlotMutationTypes}{PlotMutationTypes}{PlotMutationTypes}
%
\begin{Description}\relax
Plots a horizontal barplot of mutation types
\end{Description}
%
\begin{Usage}
\begin{verbatim}
PlotMutationTypes(mutation.types = c("C>A", "C>G", "C>T", "T>A", "T>C",
  "T>G"), mutation.types.values, mutation.types.colors, max.y.val, title,
  convert.to.percentage = T, show.legend = T, font.size.small = 8,
  font.size.med = 14, plot.margin = unit(c(0.5, 0.5, 0.5, 0.5), "cm"))
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{mutation.types}] Mutation types; Default = c("C>A", "C>G", "C>T", "T>A", "T>C", "T>G")

\item[\code{mutation.types.values}] Mutation count for each mutation type

\item[\code{mutation.types.colors}] A color vector for indicating mutation types

\item[\code{max.y.val}] y axis maximum value

\item[\code{title}] Plot title

\item[\code{convert.to.percentage}] if True convert y values to percentage (x 100); Default = T

\item[\code{show.legend}] If True, show legend; Default = T

\item[\code{font.size.small}] Small font size; Default = 8

\item[\code{font.size.med}] Medium font size; Default = 14

\item[\code{plot.margin}] Margin vector for drawing plot; Default = unit(c(0.5, 0.5, 0.5, 0.5), "cm"))
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A ggplot object
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
## Not run: 
p <- PlotMutationTypes(mutation.types = c("C>A", "C>G", "C>T", "T>A", "T>C", "T>G"),
                  mutation.types.values = c(0.3, 0.3, 0.1, 0.1, 0.1, 0.1),
                  mutation.types.colors = TriNuc.Mutation.Type.Hex.Colors,
                  max.y.val = 0.5,
                  convert.to.percentage = T,
                  show.legend = T,
                  font.size.small = 8,
                  font.size.med = 14,
                  plot.margin = unit(c(0.5, 0.5, 0.5, 0.5), "cm"))
print(p)

## End(Not run)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{PlotOptimizationIterations}{PlotOptimizationIterations}{PlotOptimizationIterations}
%
\begin{Description}\relax
Plots multiple scatter plots into one figure
\end{Description}
%
\begin{Usage}
\begin{verbatim}
PlotOptimizationIterations(df, columns.to.plot, x.axis.var, x.axis.title,
  x.min, x.max, save.file, title, y.axis.title = "", y.max = 1,
  point.size = 1, connect.dots = T, plot.legend = T,
  legend.ncol = 1, font.size.med = 14, font.size.large = 16,
  plot.margin = unit(c(0.5, 0.5, 0.5, 0.5), "cm"))
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{df}] A data.frame (from reading "FIREVAT\_Optimization\_Logs.tsv")

\item[\code{columns.to.plot}] A character vector (of column names to plot)

\item[\code{x.axis.var}] x axis variable

\item[\code{x.axis.title}] x axis title

\item[\code{x.max}] x axis maximum value

\item[\code{save.file}] Filename (including full path) to which the plot will be saved

\item[\code{title}] Plot title

\item[\code{y.axis.title}] y axis title; Default = ""

\item[\code{y.max}] y axis maximum value; Default = 1

\item[\code{point.size}] Point size; Default = 1

\item[\code{connect.dots}] If True draws dots for each iteration; Default = True

\item[\code{plot.legend}] If True write legend of plot; Default = T

\item[\code{legend.ncol}] legend.n Default = 1

\item[\code{font.size.med}] Medium font size; Default = 14

\item[\code{font.size.large}] Large font size; Default = 16

\item[\code{plot.margin}] Margin vector for plot; Default = unit(c(0.5, 0.5, 0.5, 0.5), "cm"))
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A ggplot object
\end{Value}
\inputencoding{utf8}
\HeaderA{PlotSignaturesContProbs}{PlotSignaturesContProbs}{PlotSignaturesContProbs}
%
\begin{Description}\relax
Plots a horizontal barplot of identified mutational signatures
\end{Description}
%
\begin{Usage}
\begin{verbatim}
PlotSignaturesContProbs(df.identified.mut.sigs, df.ref.sigs.groups.colors,
  title, convert.to.percentage = T, font.size.small = 8,
  font.size.med = 14, plot.margin = unit(c(0.5, 0.5, 0.5, 0.5), "cm"))
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{df.identified.mut.sigs}] A data.frame of identified mutational signatures

\item[\code{df.ref.sigs.groups.colors}] A data.frame with 'signature', 'group', and 'color' columns

\item[\code{title}] Plot title

\item[\code{convert.to.percentage}] If true, convert y values to percentage (x 100); Default = T,

\item[\code{font.size.small}] Small font size; Default = 8,

\item[\code{font.size.med}] Medium font size; Default = 14,

\item[\code{plot.margin}] Margin vector for drawing plot; Default = unit(c(0.5, 0.5, 0.5, 0.5), "cm"))
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A ggplot object
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
## Not run: 
 g <- PlotSignaturesContProbs(sigs = c(mutalisk.results$identified.mut.sigs),
 sigs.probs = c(mutalisk.results$identified.mut.sigs.probs),
 df.ref.sigs.groups.colors = GetPCAWGMutSigsEtiologiesColors())
 print(g)

## End(Not run)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{PlotTable}{PlotTable}{PlotTable}
%
\begin{Description}\relax
Plots basic statistics table
\end{Description}
%
\begin{Usage}
\begin{verbatim}
PlotTable(df, padding = 20, font.size = 14)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{df}] = A data.frame where the first column is header and the second column is data value

\item[\code{padding}] Padding size; Default = 20

\item[\code{font.size}] Font size; Default = 14
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A plot
\end{Value}
\inputencoding{utf8}
\HeaderA{PlotTriNucSpectrum}{PlotTriNucSpectrum}{PlotTriNucSpectrum}
%
\begin{Description}\relax
Plots the spectrum of 96 trinucleotide distribution
(C>A, C>G, C>T, T>A, T>C, T>G)
Please note that this function assumes that both sub.types and spectrum
are sorted in the following order: C>A, C>G, C>T, T>A, T>C, T>G
\end{Description}
%
\begin{Usage}
\begin{verbatim}
PlotTriNucSpectrum(sub.types, spectrum, max.y.val, min.y.val, y.axis.title,
  draw.top.strip = T, draw.x.axis.labels = T, draw.y.axis.labels = T,
  draw.y.axis.title = T, font.size.small = 8, font.size.med = 14,
  plot.margin.top = 0.5, plot.margin.bottom = 0.5,
  plot.margin.left = 0.5, plot.margin.right = 0.5, title)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{sub.types}] A character vector (types of 96 trinucleotide substitutions)

\item[\code{spectrum}] A numeric vector (96 elements)

\item[\code{max.y.val}] y axis maximum value

\item[\code{min.y.val}] y axis minimum value

\item[\code{y.axis.title}] y axis title

\item[\code{draw.top.strip}] If True then draws top strip; Default = T

\item[\code{draw.x.axis.labels}] If True then draws x axis labels; Default = T

\item[\code{draw.y.axis.labels}] If True then draws y axis labels; Default = T

\item[\code{draw.y.axis.title}] If True then draws y axis title; Default = T

\item[\code{font.size.small}] Small font size; Default = 8

\item[\code{font.size.med}] Medium font size; Default = 14

\item[\code{plot.margin.top}] Top margin; Default = 0.5

\item[\code{plot.margin.bottom}] Bottom margin; Default = 0.5

\item[\code{plot.margin.left}] Left margin; Default = 0.5

\item[\code{plot.margin.right}] Right margin; Default = 0.5

\item[\code{title}] Plot title
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A ggplot object
\end{Value}
\inputencoding{utf8}
\HeaderA{PlotVCFStatsBoxPlots}{PlotVCFStatsBoxPlots}{PlotVCFStatsBoxPlots}
%
\begin{Description}\relax
Plots multiple (original, refined, artifact vcf) boxplots for single filter parameter
\end{Description}
%
\begin{Usage}
\begin{verbatim}
PlotVCFStatsBoxPlots(original.vcf.stat.values, refined.vcf.stat.values,
  artifact.vcf.stat.values, xlab, axis.font.size = 10,
  label.font.size = 10, title.font.size = 12)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{original.vcf.stat.values}] A numeric vector corresponding to the original vcf.obj values of single filter parameter

\item[\code{refined.vcf.stat.values}] A numeric vector corresponding to the refined vcf.obj values of single filter parameter

\item[\code{artifact.vcf.stat.values}] A numeric vector corresponding to the artifact vcf.obj values of single filter parameter

\item[\code{xlab}] A string value (x-axis label)

\item[\code{axis.font.size}] An integer value (axis font size)

\item[\code{label.font.size}] An integer value (label font size)

\item[\code{title.font.size}] An integer value (title font size)
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A ggboxplot
\end{Value}
\inputencoding{utf8}
\HeaderA{PlotVCFStatsHistograms}{PlotVCFStatsHistograms}{PlotVCFStatsHistograms}
%
\begin{Description}\relax
Plots multiple VCF stats histograms into one figure
\end{Description}
%
\begin{Usage}
\begin{verbatim}
PlotVCFStatsHistograms(plot.values, x.axis.labels, stat.y.max.vals,
  stat.x.max.vals, sample.id, save.file, title, cutoff.values,
  plot.boxplot = F, plot.cutoff.line.color = "#D4012E",
  plot.cutoff.value.lines = F, bin.width = 1, ncol = 4, nrow = 3,
  font.size.med = 10, font.size.large = 12, plot.margin = unit(c(0.5,
  0.5, 0.5, 0.5), "cm"))
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{plot.values}] A list of multiple numeric vectors

\item[\code{x.axis.labels}] A character vector of x axis labels

\item[\code{stat.y.max.vals}] A numeric vector of max y-axis values

\item[\code{stat.x.max.vals}] A numeric vector of max x-axis values

\item[\code{sample.id}] A string value of sample ID

\item[\code{save.file}] A string value of file to which the resulting plot will be saved

\item[\code{title}] A string value of plot title

\item[\code{cutoff.values}] A numeric vector of cutoff values

\item[\code{plot.boxplot}] A boolean value (default = False)

\item[\code{plot.cutoff.line.color}] A hex string value (default = "\#D4012E")

\item[\code{plot.cutoff.value.lines}] A boolean value (default = False)

\item[\code{bin.width}] An integer value (default = 1; histogram bin width)

\item[\code{ncol}] An integer value (default = 4; ggarrange ncol)

\item[\code{nrow}] An integer value (default = 3; ggarrange nrow)

\item[\code{font.size.med}] An integer value (default = 10)

\item[\code{font.size.large}] An integer value (default = 12)

\item[\code{plot.margin}] A list (default = unit(c(0.5, 0.5, 0.5, 0.5), "cm"))
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A list with the following elements
\begin{itemize}

\item{} f = A ggarrange object
\item{} graphs = A list of length 3; each element is a ggplot histogram

\end{itemize}

\end{Value}
\inputencoding{utf8}
\HeaderA{PrepareAnnotationDB}{PrepareAnnotationDB}{PrepareAnnotationDB}
%
\begin{Description}\relax
Prepares df.genes.of.interest from a vcf.obj (\code{\LinkA{ReadVCF}{ReadVCF}})
of COSMIC or ClinVar vcf for \code{\LinkA{AnnotateVCFObj}{AnnotateVCFObj}}
\end{Description}
%
\begin{Usage}
\begin{verbatim}
PrepareAnnotationDB(annotation.vcf.obj)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{annotation.vcf.obj}] vcf.obj of COSMIC or ClinVar vcf file
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A data.frame of annotation.vcf.obj
\end{Value}
\inputencoding{utf8}
\HeaderA{PrepareArtifactAnnotationTable}{PrepareArtifactAnnotationTable}{PrepareArtifactAnnotationTable}
%
\begin{Description}\relax
Prepares artifactual mutations annotation (filtered, queried) table
\end{Description}
%
\begin{Usage}
\begin{verbatim}
PrepareArtifactAnnotationTable(data)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{data}] A list of elements returned from \code{\LinkA{RunFIREVAT}{RunFIREVAT}}
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A data.frame
\end{Value}
\inputencoding{utf8}
\HeaderA{PrepareArtifactStrandBiasTable}{PrepareArtifactStrandBiasTable}{PrepareArtifactStrandBiasTable}
%
\begin{Description}\relax
Prepares artifactual mutations strand biased variants table
\end{Description}
%
\begin{Usage}
\begin{verbatim}
PrepareArtifactStrandBiasTable(data)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{data}] A list of elements returned from \code{\LinkA{RunFIREVAT}{RunFIREVAT}}
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A data.frame
\end{Value}
\inputencoding{utf8}
\HeaderA{PrepareArtifactualMutsOptimizationIterationsPlot}{PrepareArtifactualMutsOptimizationIterationsPlot}{PrepareArtifactualMutsOptimizationIterationsPlot}
%
\begin{Description}\relax
Prepares artifactual mutations optimization iterations plot
\end{Description}
%
\begin{Usage}
\begin{verbatim}
PrepareArtifactualMutsOptimizationIterationsPlot(data)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{data}] A list of elements returned from \code{\LinkA{RunFIREVAT}{RunFIREVAT}}
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A ggplot object
\end{Value}
\inputencoding{utf8}
\HeaderA{PrepareFilterCutoffsTable}{PrepareFilterCutoffsTable}{PrepareFilterCutoffsTable}
%
\begin{Description}\relax
Prepares filter cutoffs table for reporting
\end{Description}
%
\begin{Usage}
\begin{verbatim}
PrepareFilterCutoffsTable(data)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{data}] A list of elements returned from \code{\LinkA{RunFIREVAT}{RunFIREVAT}}
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A data.frame
\end{Value}
\inputencoding{utf8}
\HeaderA{PrepareGeneticAlgorithmParametersTable}{PrepareGeneticAlgorithmParametersTable}{PrepareGeneticAlgorithmParametersTable}
%
\begin{Description}\relax
Prepares Genetic Algorithm parameters table
\end{Description}
%
\begin{Usage}
\begin{verbatim}
PrepareGeneticAlgorithmParametersTable(data)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{data}] A list of elements returned from \code{\LinkA{RunFIREVAT}{RunFIREVAT}}
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A data.frame
\end{Value}
\inputencoding{utf8}
\HeaderA{PrepareIdentifiedSignaturesPlot}{PrepareIdentifiedSignaturesPlot}{PrepareIdentifiedSignaturesPlot}
%
\begin{Description}\relax
Prepares identified signatures plot for reporting
\end{Description}
%
\begin{Usage}
\begin{verbatim}
PrepareIdentifiedSignaturesPlot(data)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{data}] A list of elements returned from \code{\LinkA{RunFIREVAT}{RunFIREVAT}}
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A ggarrange object
\end{Value}
\inputencoding{utf8}
\HeaderA{PrepareMLEReconstructedSpectrumsPlot}{PrepareMLEReconstructedSpectrumsPlot}{PrepareMLEReconstructedSpectrumsPlot}
%
\begin{Description}\relax
Prepares MLE reconstructed spectrums plot
\end{Description}
%
\begin{Usage}
\begin{verbatim}
PrepareMLEReconstructedSpectrumsPlot(data)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{data}] A list of elements returned from \code{\LinkA{RunFIREVAT}{RunFIREVAT}}
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A ggarrange object
\end{Value}
\inputencoding{utf8}
\HeaderA{PrepareNucleotideSubstitutionTypesPlot}{PrepareNucleotideSubstitutionTypesPlot}{PrepareNucleotideSubstitutionTypesPlot}
%
\begin{Description}\relax
Prepares nucleotide substitution types plot
\end{Description}
%
\begin{Usage}
\begin{verbatim}
PrepareNucleotideSubstitutionTypesPlot(data)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{data}] A list of elements returned from \code{\LinkA{RunFIREVAT}{RunFIREVAT}}
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A ggarrange object
\end{Value}
\inputencoding{utf8}
\HeaderA{PrepareObservedSpectrumsPlot}{PrepareObservedSpectrumsPlot}{PrepareObservedSpectrumsPlot}
%
\begin{Description}\relax
Prepares observed spectrums plot
\end{Description}
%
\begin{Usage}
\begin{verbatim}
PrepareObservedSpectrumsPlot(data)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{data}] A list of elements returned from \code{\LinkA{RunFIREVAT}{RunFIREVAT}}
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A ggarrange object
\end{Value}
\inputencoding{utf8}
\HeaderA{PrepareOptimizationResultsTable}{PrepareOptimizationResultsTable}{PrepareOptimizationResultsTable}
%
\begin{Description}\relax
Prepares optimization results table
\end{Description}
%
\begin{Usage}
\begin{verbatim}
PrepareOptimizationResultsTable(data)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{data}] A list of elements returned from \code{\LinkA{RunFIREVAT}{RunFIREVAT}}
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A data.frame
\end{Value}
\inputencoding{utf8}
\HeaderA{PrepareOptimizedVCFStatisticsPlot}{PrepareOptimizedVCFStatisticsPlot}{PrepareOptimizedVCFStatisticsPlot}
%
\begin{Description}\relax
Prepares optimized VCF statistics plot
\end{Description}
%
\begin{Usage}
\begin{verbatim}
PrepareOptimizedVCFStatisticsPlot(data)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{data}] A list of elements returned from \code{\LinkA{RunFIREVAT}{RunFIREVAT}}
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A ggarrange object
\end{Value}
\inputencoding{utf8}
\HeaderA{PrepareRefinedAnnotationTable}{PrepareRefinedAnnotationTable}{PrepareRefinedAnnotationTable}
%
\begin{Description}\relax
Prepares refined mutations annotation (filtered, queried) table
\end{Description}
%
\begin{Usage}
\begin{verbatim}
PrepareRefinedAnnotationTable(data)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{data}] A list of elements returned from \code{\LinkA{RunFIREVAT}{RunFIREVAT}}
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A data.frame
\end{Value}
\inputencoding{utf8}
\HeaderA{PrepareRefinedMutsOptimizationIterationsPlot}{PrepareRefinedMutsOptimizationIterationsPlot}{PrepareRefinedMutsOptimizationIterationsPlot}
%
\begin{Description}\relax
Prepares refined mutations optimization iterations plot
\end{Description}
%
\begin{Usage}
\begin{verbatim}
PrepareRefinedMutsOptimizationIterationsPlot(data)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{data}] A list of elements returned from \code{\LinkA{RunFIREVAT}{RunFIREVAT}}
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A ggplot object
\end{Value}
\inputencoding{utf8}
\HeaderA{PrepareRefinedStrandBiasTable}{PrepareRefinedStrandBiasTable}{PrepareRefinedStrandBiasTable}
%
\begin{Description}\relax
Prepares refined mutations strand biased variants table
\end{Description}
%
\begin{Usage}
\begin{verbatim}
PrepareRefinedStrandBiasTable(data)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{data}] A list of elements returned from \code{\LinkA{RunFIREVAT}{RunFIREVAT}}
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A data.frame
\end{Value}
\inputencoding{utf8}
\HeaderA{PrepareResidualSpectrumsPlot}{PrepareResidualSpectrumsPlot}{PrepareResidualSpectrumsPlot}
%
\begin{Description}\relax
Prepares residual spectrums plot
\end{Description}
%
\begin{Usage}
\begin{verbatim}
PrepareResidualSpectrumsPlot(data)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{data}] A list of elements returned from \code{\LinkA{RunFIREVAT}{RunFIREVAT}}
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A ggarrange object
\end{Value}
\inputencoding{utf8}
\HeaderA{PrepareTrinucleotideSpectrumsTable}{PrepareTrinucleotideSpectrumsTable}{PrepareTrinucleotideSpectrumsTable}
%
\begin{Description}\relax
Prepares trinucleotide spectrums table
\end{Description}
%
\begin{Usage}
\begin{verbatim}
PrepareTrinucleotideSpectrumsTable(data)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{data}] A list of elements returned from \code{\LinkA{RunFIREVAT}{RunFIREVAT}}
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A data.frame
\end{Value}
\inputencoding{utf8}
\HeaderA{PrintLog}{PrintLog}{PrintLog}
%
\begin{Description}\relax
Prints log message
\end{Description}
%
\begin{Usage}
\begin{verbatim}
PrintLog(msg, type = "INFO")
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{msg}] String value message to print along with log type and date

\item[\code{type}] String value that represents type of this message. 'INFO' by default.
\end{ldescription}
\end{Arguments}
\inputencoding{utf8}
\HeaderA{QueryAnnotatedVCF}{FilterAnnotatedVCF}{QueryAnnotatedVCF}
%
\begin{Description}\relax
Annotates a vcf.obj using df.variants.of.interest (from (\code{\LinkA{PrepareAnnotationDB}{PrepareAnnotationDB}})
\end{Description}
%
\begin{Usage}
\begin{verbatim}
QueryAnnotatedVCF(vcf.obj.annotated, filter.key.value.pairs,
  filter.condition = "AND")
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{vcf.obj.annotated}] \code{\LinkA{AnnotateVCFObj}{AnnotateVCFObj}}

\item[\code{filter.key.value.pairs}] A list with the key as the column name and value as the
filtering values. E.g. list("CLNSIG" = c("Pathogenic", "Pathogenic/Likely\_pathogenic"))

\item[\code{filter.condition}] 'AND' or 'OR'.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A vcf.obj
\end{Value}
\inputencoding{utf8}
\HeaderA{ReadOptimizationIterationReport}{ReadOptimizationIterationReport}{ReadOptimizationIterationReport}
%
\begin{Description}\relax
Read optimization iteration report
\end{Description}
%
\begin{Usage}
\begin{verbatim}
ReadOptimizationIterationReport(data)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{data}] A list of elements returned from \code{\LinkA{RunFIREVAT}{RunFIREVAT}}
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A data.frame of FIREVAT optimization logs
\end{Value}
\inputencoding{utf8}
\HeaderA{ReadVCF}{ReadVCF}{ReadVCF}
%
\begin{Description}\relax
Reads a .vcf file
\end{Description}
%
\begin{Usage}
\begin{verbatim}
ReadVCF(vcf.file, genome = "hg19", split.info = FALSE,
  check.chromosome.name = TRUE)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{vcf.file}] (full path of a .vcf file)

\item[\code{genome}] ('hg19' or 'hg38')

\item[\code{split.info}] A boolean value. If TRUE, then makes the INFO column in the vcf
as a separate column. Default value is FALSE.

\item[\code{check.chromosome.name}] A boolean value. If TRUE, then check whether converts
'MT' to 'M' and adds 'chr' to the CHROM column. Default value is TRUE.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A list with elements 'data', 'header', 'genome'
\end{Value}
\inputencoding{utf8}
\HeaderA{ReportFIREVATResults}{ReportFIREVATResults}{ReportFIREVATResults}
%
\begin{Description}\relax
Reports FIREVAT results in html format (generated from Rmd)
\end{Description}
%
\begin{Usage}
\begin{verbatim}
ReportFIREVATResults(data)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{data}] A list of main data from \code{\LinkA{RunFIREVAT}{RunFIREVAT}}
\end{ldescription}
\end{Arguments}
%
\begin{Value}
An updated data list
\end{Value}
\inputencoding{utf8}
\HeaderA{RunFIREVAT}{RunFIREVAT}{RunFIREVAT}
%
\begin{Description}\relax
Runs FIREVAT using configuration data. Filters point mutations in the user-specified vcf file based on mutational signature
identification and outputs the refined and artifact vcf files as well as metadata related to the refinement process.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
RunFIREVAT(vcf.file, vcf.file.genome = "hg19", config.file,
  df.ref.mut.sigs = GetPCAWGMutSigs(),
  target.mut.sigs = GetPCAWGMutSigsNames(),
  df.ref.mut.sigs.groups.colors = GetPCAWGMutSigsEtiologiesColors(),
  sequencing.artifact.mut.sigs = PCAWG.All.Sequencing.Artifact.Signatures,
  num.cores = 2, output.dir, mode = "ga", init.artifact.stop = 0.05,
  objective.fn = Default.Obj.Fn, use.suggested.soln = TRUE,
  ga.type = "real-valued", ga.pop.size = 100, ga.max.iter = 100,
  ga.run = 50, ga.pmutation = 0.1, ga.preemptive.killing = FALSE,
  ga.seed = NULL, mutalisk = TRUE, mutalisk.method = "all",
  mutalisk.must.include.sigs = NULL,
  mutalisk.random.sampling.count = 20,
  mutalisk.random.sampling.max.iter = 10,
  perform.strand.bias.analysis = FALSE,
  filter.by.strand.bias.analysis = TRUE,
  filter.by.strand.bias.analysis.cutoff = 0.25,
  strand.bias.perform.fdr.correction = TRUE,
  strand.bias.fdr.correction.method = "BH",
  ref.forward.strand.var = NULL, ref.reverse.strand.var = NULL,
  alt.forward.strand.var = NULL, alt.reverse.strand.var = NULL,
  annotate = FALSE, df.annotation.db = NULL,
  annotated.columns.to.display = NULL,
  annotation.filter.key.value.pairs = NULL,
  annotation.filter.condition = "AND", write.vcf = TRUE,
  report = TRUE, save.rdata = TRUE, save.tsv = TRUE,
  report.format = "html", verbose = TRUE)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{vcf.file}] String value corresponding to input .vcf file. Please provide the full path.

\item[\code{vcf.file.genome}] Genome assembly of the input .vcf file. The value should be eitehr 'hg19' or 'hg38'.

\item[\code{config.file}] String value corresponding to input configuration file. For more details please refer to ...

\item[\code{df.ref.mut.sigs}] A data.frame of the reference mutational signatures

\item[\code{target.mut.sigs}] A character vector of the target mutational signatures from reference mutational signatures.

\item[\code{df.ref.mut.sigs.groups.colors}] A data.frame of the reference mutational signatures groups and colors

\item[\code{sequencing.artifact.mut.sigs}] A character vector of the sequencing artifact mutational signatures from reference mutational signatures.

\item[\code{num.cores}] Number of cores to allocate

\item[\code{output.dir}] String value of the desired output directory

\item[\code{mode}] String value. The value should be either 'ga' or 'manual'.

\item[\code{init.artifact.stop}] Numeric value  less than 1. If the sum of sequencing artifact weights in the user-specified original VCF file (i.e. vcf.file)
is less than or equal to this value then FIREVAT does not perform variant refinement.
Default value is 0.05. Note that this option does not apply if 'mode' is 'manual'.

\item[\code{objective.fn}] Objective value derivation function. Default: Default.Obj.Fn.

\item[\code{use.suggested.soln}] Boolean value. If TRUE, then FIREVAT passes the default values
of filter variables declared as 'use\_in\_filter' in the config file to the 'suggestions' parameter of
the Genetic Algorithm package. If FALSE, then FIREVAT supplies NULL to the GA package 'suggestions' parameter.
FIREVAT also computes baseline performance of each filter variable and uses fittest population from each variable
as a suggested solution.

\item[\code{ga.type}] String value. The value should be either 'binray' or 'real-valued'.

\item[\code{ga.pop.size}] Integer value of the Genetic Algorithm 'population size' parameter. Default: 100.
This value should be set based on the number of filter parameters. Recommendation: 40 per filter parameter.

\item[\code{ga.max.iter}] Integer value of the Genetic Algorithm 'maximum iterations' parameter. Default: 100.
This value should be set based on the number of filter parameters. Recommendation: same as 'ga.pop.size'.

\item[\code{ga.run}] Integer value of the Genetic Algorithm 'run' parameter. Default: 50.
This value should be set based on the 'ga.max.iter' parameter. Recommendation: 25 percent of 'ga.max.iter'.

\item[\code{ga.pmutation}] Float value of the Genetic Algorithm 'mutation probability' parameter. Default: 0.1.

\item[\code{ga.preemptive.killing}] If TRUE, then preemptively kills populations that yield greater sequencing artifact weights sum
compared to the original mutatational signatures analysis

\item[\code{ga.seed}] Integer value of the Genetic Algorithm 'seed' parameter. Default: NULL.

\item[\code{mutalisk}] If TRUE, confirm mutational signature analysis with Mutalisk. Default: TRUE.

\item[\code{mutalisk.method}] Mutalisk signature identification method. Default: 'random.sampling'.
The value can be either 'all' or 'random.sampling'.
'all' uses all target.mut.sigs to identify mutational signatures.
'random.sampling' randomly samples from target.mut.sigs to identify mutational signatures.

\item[\code{mutalisk.must.include.sigs}] Signatures that must be included in the Mutalisk signature identification
A character vector corresponding to the signature names.

\item[\code{mutalisk.random.sampling.count}] Mutalisk random sampling count. Default: 20.
The number of signatures to sample from target.mut.sigs

\item[\code{mutalisk.random.sampling.max.iter}] Mutalisk random sampling maximum iteration. Default: 10.
The number of times Mutalisk randomly samples from target.mut.sigs before determining the candidate signatures.

\item[\code{perform.strand.bias.analysis}] If TRUE, then performs strand bias analysis.

\item[\code{filter.by.strand.bias.analysis}] If TRUE, then filters out variants in refined vcf based on strand bias analysis results

\item[\code{filter.by.strand.bias.analysis.cutoff}] The p.value or q value cutoff for filtering out variants.

\item[\code{strand.bias.perform.fdr.correction}] If TRUE, then performs false discovery rate
correction for strand bias analysis.

\item[\code{strand.bias.fdr.correction.method}] A string value. Default value is 'BH'.
Refer to 'p.adjust()' function method.

\item[\code{ref.forward.strand.var}] A string value.

\item[\code{ref.reverse.strand.var}] A string value,

\item[\code{alt.forward.strand.var}] A string value,

\item[\code{alt.reverse.strand.var}] A string value,

\item[\code{annotate}] A boolean value. Default value is TRUE.

\item[\code{df.annotation.db}] A data.frame. Please refer to \code{\LinkA{PrepareAnnotationDB}{PrepareAnnotationDB}}

\item[\code{annotated.columns.to.display}] A character vector.

\item[\code{annotation.filter.key.value.pairs}] A list.

\item[\code{annotation.filter.condition}] 'AND' or 'OR'.

\item[\code{write.vcf}] If TRUE, write original/refined/artifact vcfs. Default: TRUE.

\item[\code{report}] If TRUE, generate report. Default: TRUE.

\item[\code{save.rdata}] If TRUE, save rdata. Default: TRUE.

\item[\code{save.tsv}] If TRUE, save tsv. Default: TRUE.

\item[\code{report.format}] The format of FIREVAT report. We currently only support 'html'.

\item[\code{verbose}] If TRUE, provides process detail. Default: TRUE.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A list with the following elements
\begin{itemize}

\item{} f = A ggarrange object
\item{} graphs = A list of length 3; each element is a ggplot histogram

\end{itemize}

\end{Value}
\inputencoding{utf8}
\HeaderA{RunGAMode}{RunGAMode}{RunGAMode}
%
\begin{Description}\relax
Runs FIREVAT ga mode
\end{Description}
%
\begin{Usage}
\begin{verbatim}
RunGAMode(data)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{data}] A list from RunFIREVAT
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A list
\end{Value}
\inputencoding{utf8}
\HeaderA{RunManualMode}{RunManualMode}{RunManualMode}
%
\begin{Description}\relax
Runs FIREVAT manual mode
\end{Description}
%
\begin{Usage}
\begin{verbatim}
RunManualMode(data)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{data}] A list from RunFIREVAT
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A list
\end{Value}
\inputencoding{utf8}
\HeaderA{RunMutalisk}{RunMutalisk}{RunMutalisk}
%
\begin{Description}\relax
Identifies mutational signatures using Mutalisk
\end{Description}
%
\begin{Usage}
\begin{verbatim}
RunMutalisk(vcf.obj, df.ref.mut.sigs, target.mut.sigs,
  random.sampling.candidate.mut.sigs = c(), method = "random.sampling",
  n.sample = 20, n.iter = 10, verbose = TRUE)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{vcf.obj}] A list (from firevat\_vcf::ReadVCF)

\item[\code{df.ref.mut.sigs}] A data.frame of reference mutational signatures

\item[\code{target.mut.sigs}] A character vector of target mutational signatures names to identify from

\item[\code{random.sampling.candidate.mut.sigs}] A character vector of mutational signatures names
that gets appended to the list of candidate mutational signatures so that these are
always considered.

\item[\code{method}] A string value (must be either 'random.sampling' or 'all').
The method 'random.sampling' samples (without replacement) 'n.sample' number of signatures
'n.iter' number of times and runs the candidate signatures one last time.
The method 'all' uses all target.mut.sigs

\item[\code{n.sample}] An integer value ('random.sampling' method parameter)
Number of signatures to choose for each iteration of random sampling).

\item[\code{n.iter}] An integer value ('random.sampling' method parameter).
Number of iterations to perform random sampling.

\item[\code{verbose}] If true, provides process details
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A list with the following elements
\begin{itemize}

\item{} num.point.mutationsAn integer value - count of total point mutations
\item{} sub.typesA character vector of length 96
\item{} sub.types.spectrumA numeric vector of length 96
\item{} num.mut.sigsAn integer value (count of unique mutational signatures identified)
\item{} identified.mut.sigsA character vector where each element is a mutational signature identified
\item{} identified.mut.sigs.probsA numeric vector where each element is the weight of mutational signature identified.
The ordering follows identified.mut.sigs
\item{} identified.mut.sigs.spectrumA numeric vector of length 96
\item{} residualsA numeric vector of length 96
\item{} rssA numeric value (residual sum of squares)
\item{} cos.sim.scoreA numeric value (cosine similarity score between observed mutational spectrum and
reconstructed mutational signatures)
\item{} all.models.sigsA list where each element is a model; a model is a list of signatures identified)
\item{} all.models.sigs.probsA list where each element is a model; a model is a list of contribution probabilities
\item{} all.models.cos.sim.scoresA list where each element is a model; a model is a list of cosine similarity socres

\end{itemize}

\end{Value}
\inputencoding{utf8}
\HeaderA{RunMutaliskHelper}{RunMutaliskHelper}{RunMutaliskHelper}
%
\begin{Description}\relax
Helper function for RunMutalisk
\end{Description}
%
\begin{Usage}
\begin{verbatim}
RunMutaliskHelper(vcf.trinucleotide.data, df.ref.mut.sigs, target.mut.sigs)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{vcf.trinucleotide.data}] A data.frame (from firevat\_mutalisk::MutaliskParseVCFObj)

\item[\code{df.ref.mut.sigs}] A data.frame of reference mutational signatures

\item[\code{target.mut.sigs}] A character vector of target mutational signatures names
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A list with the following elements
\begin{itemize}

\item{} num.point.mutationsAn integer value - count of total point mutations
\item{} sub.typesA character vector of length 96
\item{} sub.types.spectrumA numeric vector of length 96
\item{} num.mut.sigsAn integer value (count of unique mutational signatures identified)
\item{} identified.mut.sigsA character vector where each element is a mutational signature identified
\item{} identified.mut.sigs.probsA numeric vector where each element is the weight of mutational signature identified.
The ordering follows identified.mut.sigs
\item{} identified.mut.sigs.spectrumA numeric vector of length 96
\item{} residualsA numeric vector of length 96
\item{} rssA numeric value (residual sum of squares)
\item{} cos.sim.scoreA numeric value (cosine similarity score between observed mutational spectrum and
reconstructed mutational signatures)
\item{} all.models.sigsA list where each element is a model; a model is a list of signatures identified)
\item{} all.models.sigs.probsA list where each element is a model; a model is a list of contribution probabilities
\item{} all.models.cos.sim.scoresA list where each element is a model; a model is a list of cosine similarity socres

\end{itemize}

\end{Value}
\inputencoding{utf8}
\HeaderA{RunMutPat}{RunMutPat}{RunMutPat}
%
\begin{Description}\relax
Identifies mutational signatures using Mutational Patterns
\end{Description}
%
\begin{Usage}
\begin{verbatim}
RunMutPat(mut.pat.input, df.mut.pat.ref.sigs, target.mut.sigs,
  verbose = TRUE)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{mut.pat.input}] A list from \code{\LinkA{MutPatParseVCFObj}{MutPatParseVCFObj}}

\item[\code{df.mut.pat.ref.sigs}] A data.frame returned by \code{\LinkA{MutPatParseRefMutSigs}{MutPatParseRefMutSigs}}

\item[\code{target.mut.sigs}] A character vector of target mutational signatures names

\item[\code{verbose}] If true, provides process details
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A list with the following elements
\begin{itemize}

\item{} tumor.mutation.types.spectrumA numeric vector of length 96 - 'observed' spectrum
\item{} identified.mutation.types.spectrumA numeric vector of length 96 - 'identified' spectrum
\item{} residualsA numeric vector of length 96 - residuals
\item{} mutation.typesA character vector of length 96
\item{} identified.mut.sigsA character vector where each element is a mutational signature identified
\item{} identified.mut.sigs.contribution.weightsA numeric vector where each element is the weight of mutational signature identified. The ordering follows identified.mut.sigs
\item{} cosine.similarity.scoreA numeric value

\end{itemize}

\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
## Not run: 
vcf.obj <- ReadVCF(vcf.file = "../data/sample/HNT-082-BT.final.call.vcf", genome = "hg19")
df.ref.mut.sigs <- GetPCAWGMutSigs()
target.mut.sigs <- GetPCAWGMutSigsNames()
RunMutPat(vcf.obj = vcf.obj,
df.ref.mut.sigs = df.ref.mut.sigs,
target.mut.sigs = target.mut.sigs)

## End(Not run)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{Sigmoid.Obj.Fn}{Sigmoid.Obj.Fn}{Sigmoid.Obj.Fn}
%
\begin{Description}\relax
Sigmoid objective function
\end{Description}
%
\begin{Usage}
\begin{verbatim}
Sigmoid.Obj.Fn(C.refined, A.refined, C.artifactual, A.artifactual)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{C.refined}] A numeric value between 0 and 1.

\item[\code{A.refined}] A numeric value between 0 and 1.

\item[\code{C.artifactual}] A numeric value between 0 and 1.

\item[\code{A.artifactual}] A numeric value between 0 and 1.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A numeric value between 0 and 1.
\end{Value}
\inputencoding{utf8}
\HeaderA{Test.Obj.Fn.1}{Test.Obj.Fn.1}{Test.Obj.Fn.1}
%
\begin{Description}\relax
Test objective function 1
\end{Description}
%
\begin{Usage}
\begin{verbatim}
Test.Obj.Fn.1(C.refined, A.refined, C.artifactual, A.artifactual)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{C.refined}] A numeric value between 0 and 1.

\item[\code{A.refined}] A numeric value between 0 and 1.

\item[\code{C.artifactual}] A numeric value between 0 and 1.

\item[\code{A.artifactual}] A numeric value between 0 and 1.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A numeric value between 0 and 1.
\end{Value}
\inputencoding{utf8}
\HeaderA{Test.Obj.Fn.2}{Test.Obj.Fn.2}{Test.Obj.Fn.2}
%
\begin{Description}\relax
Test objective function 2
\end{Description}
%
\begin{Usage}
\begin{verbatim}
Test.Obj.Fn.2(C.refined, A.refined, C.artifactual, A.artifactual)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{C.refined}] A numeric value between 0 and 1.

\item[\code{A.refined}] A numeric value between 0 and 1.

\item[\code{C.artifactual}] A numeric value between 0 and 1.

\item[\code{A.artifactual}] A numeric value between 0 and 1.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A numeric value between 0 and 1.
\end{Value}
\inputencoding{utf8}
\HeaderA{TriNuc.Mutation.Type.Hex.Colors}{Constant}{TriNuc.Mutation.Type.Hex.Colors}
\keyword{datasets}{TriNuc.Mutation.Type.Hex.Colors}
%
\begin{Description}\relax
Hex codes for the mutation types (for plotting purposes)
\end{Description}
%
\begin{Usage}
\begin{verbatim}
TriNuc.Mutation.Type.Hex.Colors
\end{verbatim}
\end{Usage}
%
\begin{Format}
An object of class \code{character} of length 6.
\end{Format}
\inputencoding{utf8}
\HeaderA{UpdateFilter}{UpdateFilter}{UpdateFilter}
%
\begin{Description}\relax
Update filter based on optim parameter values
\end{Description}
%
\begin{Usage}
\begin{verbatim}
UpdateFilter(vcf.filter, param.values)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{vcf.filter}] A list from MakeFilterFromConfig

\item[\code{param.values}] A numeric vector contains filtering value
(same length with length(vcf.config.filter))
\end{ldescription}
\end{Arguments}
%
\begin{Value}
Updated vcf.filter (list)
\end{Value}
\inputencoding{utf8}
\HeaderA{WriteFIREVATResultsToTSV}{WriteFIREVATResultsToTSV}{WriteFIREVATResultsToTSV}
%
\begin{Description}\relax
Writes FIREVAT results to a csv file
\end{Description}
%
\begin{Usage}
\begin{verbatim}
WriteFIREVATResultsToTSV(firevat.results)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{firevat.results}] List returned from RunFIREVAT
\end{ldescription}
\end{Arguments}
\inputencoding{utf8}
\HeaderA{WriteVCF}{WriteVCF}{WriteVCF}
%
\begin{Description}\relax
Writes a vcf.obj to a .vcf file
\end{Description}
%
\begin{Usage}
\begin{verbatim}
WriteVCF(vcf.obj, save.file)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{vcf.obj}] (from the function ReadVCF)

\item[\code{save.file}] (full path including filename)
\end{ldescription}
\end{Arguments}
\printindex{}
\end{document}
